--- a/dwl.c
+++ b/dwl.c
@@ -299,6 +299,7 @@
 static void locksession(struct wl_listener *listener, void *data);
 static void mapnotify(struct wl_listener *listener, void *data);
 static void maximizenotify(struct wl_listener *listener, void *data);
+static struct wlr_box hideedgeborders(Client *c, struct wlr_box geo);
 static void monocle(Monitor *m);
 static void motionabsolute(struct wl_listener *listener, void *data);
 static void motionnotify(uint32_t time, struct wlr_input_device *device, double sx,
@@ -1511,6 +1512,29 @@
 	focusclient(c, 1);
 }
 
+struct wlr_box
+hideedgeborders(Client *c, struct wlr_box geo)
+{
+	struct wlr_box *w = &c->mon->w;
+	int bw = c->bw;
+
+	if (!bw)
+		return geo;
+	if (geo.x == w->x) {
+		geo.x -= bw;
+		geo.width += bw;
+	}
+	if (geo.y == w->y) {
+		geo.y -= bw;
+		geo.height += bw;
+	}
+	if (geo.x + geo.width == w->x + w->width)
+		geo.width += bw;
+	if (geo.y + geo.height == w->y + w->height)
+		geo.height += bw;
+	return geo;
+}
+
 /* We probably should change the name of this: it sounds like it
  * will focus the topmost client of this mon, when actually will
  * only return that client */
@@ -1958,11 +1982,10 @@
 	if (cursor_mode != CurNormal && cursor_mode != CurPressed)
 		return;
 	xytonode(cursor->x, cursor->y, NULL, &grabc, NULL, NULL, NULL);
-	if (!grabc || client_is_unmanaged(grabc) || grabc->isfullscreen)
+	if (!grabc || client_is_unmanaged(grabc) || grabc->isfullscreen || !grabc->isfloating)
 		return;
 
-	/* Float the window and tell motionnotify to grab it */
-	setfloating(grabc, 1);
+	/* Tell motionnotify to grab a floating window */
 	switch (cursor_mode = arg->ui) {
 	case CurMove:
 		grabcx = (int)round(cursor->x) - grabc->geom.x;
@@ -2715,6 +2738,7 @@
 	unsigned int mw, my, ty;
 	int i, n = 0;
 	Client *c;
+	struct wlr_box geo;
 
 	wl_list_for_each(c, &clients, link)
 		if (VISIBLEON(c, m) && !c->isfloating && !c->isfullscreen)
@@ -2731,13 +2755,15 @@
 		if (!VISIBLEON(c, m) || c->isfloating || c->isfullscreen)
 			continue;
 		if (i < m->nmaster) {
-			resize(c, (struct wlr_box){.x = m->w.x, .y = m->w.y + my, .width = mw,
-				.height = (m->w.height - my) / (MIN(n, m->nmaster) - i)}, 0);
-			my += c->geom.height;
+			geo = (struct wlr_box){.x = m->w.x, .y = m->w.y + my, .width = mw,
+				.height = (m->w.height - my) / (MIN(n, m->nmaster) - i)};
+			resize(c, hideedgeborders(c, geo), 0);
+			my += geo.height;
 		} else {
-			resize(c, (struct wlr_box){.x = m->w.x + mw, .y = m->w.y + ty,
-				.width = m->w.width - mw, .height = (m->w.height - ty) / (n - i)}, 0);
-			ty += c->geom.height;
+			geo = (struct wlr_box){.x = m->w.x + mw, .y = m->w.y + ty,
+				.width = m->w.width - mw, .height = (m->w.height - ty) / (n - i)};
+			resize(c, hideedgeborders(c, geo), 0);
+			ty += geo.height;
 		}
 		i++;
 	}
